core/src/main/java/org/jdbi/v3/core/mapper/GenericMapMapperFactory.java

```diff
@@ -0,0 +1,136 @@
+        }
+        Type valueType = maybeValueType.get();
+
+        // ... that is not Object...
+        if (valueType == Object.class) {
+            return Optional.empty();
+        }
+        Optional<ColumnMapper<?>> maybeColumnMapper = config.get(ColumnMappers.class).findFor(valueType);
+
+        // ... for which there is a ColumnMapper
+        return maybeColumnMapper.map(t -> forValueType(valueType));
+    }
+
+    @Beta
+    public static <X> RowMapper<Map<String, X>> forValueType(Type valueType) {
+        return new GenericMapMapper<>(valueType);
+    }
+
+    private static class GenericMapMapper<X> implements RowMapper<Map<String, X>> {
+        private final Type type;
+
+        private GenericMapMapper(Type type) {
+            this.type = type;
+        }
+
+        @Override
+        public Map<String, X> map(ResultSet rs, StatementContext ctx) throws SQLException {
+            return specialize(rs, ctx).map(rs, ctx);
+        }
+
+        @Override
+        public RowMapper<Map<String, X>> specialize(ResultSet rs, StatementContext ctx) throws SQLException {
+            Optional<ColumnMapper<?>> maybeColumnMapper = ctx.getConfig(ColumnMappers.class).findFor(type);
+
+            // unfortunately this cannot be checked sooner
+            if (maybeColumnMapper.isPresent()) {
+                ColumnMapper<?> columnMapper = maybeColumnMapper.get();
+                List<String> columnNames = getColumnNames(rs.getMetaData(), ctx.getConfig(MapMappers.class).getCaseChange());
+
+                return (r, c) -> {
+                    Map<String, X> row = new HashMap<>();
+
+                    for (int i = 0; i < columnNames.size(); i++) {
+                        @SuppressWarnings("unchecked")
+                        X value = (X) columnMapper.map(r, i + 1, ctx);
+                        row.put(columnNames.get(i), value);
+                    }
+
+                    return row;
+                };
+            } else {
+                throw new RuntimeException("no mapper found for type " + type.getTypeName());
+            }
+        }
+
+        private static List<String> getColumnNames(ResultSetMetaData meta, UnaryOperator<String> caseChange) throws SQLException {
+            // important: ordered, not sorted, and unique
+            Set<String> names = new LinkedHashSet<>();
+            int columnCount = meta.getColumnCount();
+
+            for (int i = 0; i < columnCount; i++) {
+                String columnName = meta.getColumnName(i + 1);
+                String columnLabel = meta.getColumnLabel(i + 1);
+
+                String name = columnLabel == null ? columnName : columnLabel;
+                name = caseChange.apply(name);
+
+                boolean added = names.add(name);
+                if (!added) {
+                    throw new RuntimeException("column \"" + name + "\" appears twice in this resultset!");
+                }
+            }
+
+            return new ArrayList<>(names);
```

Comment:
You could avoid a list copy here if you return it as just a `Collection` and iterate over it; you already built it as a linked list.

Reply:
need to iterate and `get` by index though

Reply:
Yes, it is a little weird.  But so is an unnecessary copy :)